src/main/java/com/flippingutilities/FlippingPlugin.java

```diff
@@ -1,609 +1,658 @@
+	 * events/duplicate events
+	 *
+	 * @return a boolean representing whether the offer should be passed on or discarded
+	 */
+	public boolean isBadEvent(GrandExchangeOfferChanged newOfferEvent)
+	{
+		GrandExchangeOffer newOffer = newOfferEvent.getOffer();
+		int newOfferSlot = newOfferEvent.getSlot();
+
+		//Check for login screen and empty offers.
+		if (newOffer.getQuantitySold() == 0 || newOfferEvent.getOffer().getItemId() == 0
+			|| client.getWidget(WidgetInfo.LOGIN_CLICK_TO_PLAY_SCREEN) != null)
+		{
+			return true;
+		}
+
+		//if its the last selling/buying event, as evidenced by the quantity sold/bought being
+		//equal to the total quantity of the offer, record it but return true so it doesn't go through
+		//as the next event will be a BOUGHT/SOLD event, and only that should go through.
+		if ((newOffer.getState() == GrandExchangeOfferState.BUYING
+			|| newOffer.getState() == GrandExchangeOfferState.SELLING)
+			&& newOffer.getQuantitySold() == newOffer.getTotalQuantity())
+		{
+			lastOffers.put(newOfferSlot, newOffer);
+			return true;
+		}
+
+		//if there is a last seen offer for that slot
+		if (lastOffers.containsKey(newOfferSlot))
+		{
+
+			GrandExchangeOffer lastOfferForSlot = lastOffers.get(newOfferSlot);
+
+			//if its a duplicate as the last seen event
+			if (lastOfferForSlot.getState().equals(newOffer.getState())
+				&& lastOfferForSlot.getQuantitySold() == newOffer.getQuantitySold())
+			{
+				return true; //its a bad event!
+			}
+			else
+			{
+				//update hashmap to include latest offer
+				lastOffers.put(newOfferSlot, newOffer);
+				return false; //not a bad event
+			}
+		}
+		//if there isn't a last seen offer for that slot
+		else
+		{
+			//put the offer in the hashmap with the corresponding slot
+			lastOffers.put(newOfferSlot, newOffer);
+			return false;
+		}
+	}
+
+	//When flipping via margin checking, we look for the lowest instant buy price
+	// to determine what our sell price should be to undercut existing offers, and vice versa for our buy price.
+	@Subscribe
+	public void onGrandExchangeOfferChanged(GrandExchangeOfferChanged newOfferEvent)
+	{
+
+		if (isBadEvent(newOfferEvent))
+		{
+			return;
+		}
+
+		final GrandExchangeOffer newOffer = newOfferEvent.getOffer();
+		final GrandExchangeOfferState newOfferState = newOffer.getState();
+
+		//Offer is a margin check.
+		//May change this in the future to be able to update prices independent of quantity.
+		//Perhaps some way of timing the state change from buying to bought such that we know it was "instantly" bought/sold?
+		if ((newOfferState == GrandExchangeOfferState.BOUGHT
+			|| newOfferState == GrandExchangeOfferState.SOLD) && newOffer.getQuantitySold() == 1)
+		{
+			addFlipTrade(tradeConstructor(newOffer));
+			panel.rebuildFlippingPanel(tradesList);
+		}
+		//If the new offer is of state BOUGHT Record the trade to keep track of GE limit.
+		else if (newOffer.getQuantitySold() > 0 && newOfferState == GrandExchangeOfferState.BOUGHT)
```

Comment:
I think this change might've been the cause for the loss of cancelled buy trades going through. Any objections to amend the condition with an ``` || newOfferState == GrandExchangeOfferState.CANCELLED_BUY)```?

Reply:
ahhh yes that makes sense. Good catch! Yeah that change should be good. I've been working on some code that delegates the responsibility of managing trade limits and refresh times to the history manager, its nice and concise due to access to standardized trades and it fixes a potential bug. Just gonna write some comments and submit a PR soon.